worker-queue-rabbit/src/main/java/com/github/workerframework/queues/rabbit/WorkerPublisherImpl.java

```
@@ -66,18 +82,52 @@ public void handlePublish(byte[] data, String routingKey, RabbitTaskInformation
-            builder.headers(headers);
+            builder.headers(publishHeaders);
             builder.contentType("text/plain");
             builder.deliveryMode(2);
-
+            
             confirmListener.registerResponseSequence(channel.getNextPublishSeqNo(), taskInformation);
-            channel.basicPublish("", routingKey, builder.build(), data);
+            channel.basicPublish("", routingKey, builder.build(), outboundByteArray);
             metrics.incrementPublished();
-        } catch (IOException e) {
+        } catch (final IOException | QueueException e) {
             LOG.error("Failed to publish result of message {} to queue {}, rejecting", taskInformation.getInboundMessageId(), routingKey, e);
             metrics.incremementErrors();
             consumerEvents.add(new ConsumerRejectEvent(Long.valueOf(taskInformation.getInboundMessageId())));
         }
     }
+
+    private boolean shouldStoreTaskMessage(final int taskMessageSize) {
+        return config.getIsMinimizationEnabled() && taskMessageSize > config.getMinimizationThreshold();
+    }
+
+    private byte[] getOutboundByteArray(
+        final byte[] taskMessage,
+        final Optional<String> taskMessagePartialRef,
+        final Map<String, Object> headers
+    ) throws QueueException {
+        try {
+            if (taskMessagePartialRef.isPresent() && shouldStoreTaskMessage(taskMessage.length)) {                
+                final var taskMessageStorageRef = dataStore.store(taskMessage, taskMessagePartialRef.get());
+                headers.put(RABBIT_HEADER_CAF_MINIMIZATION_ID, taskMessageStorageRef);
+                //  if the header is set, the consumer will ignore the incoming byte[] and use the minimized message.
+                return new byte[0];
+            }
+            return taskMessage;
+        } catch (final Exception e) {
+            throw new QueueException("Error minimizing task message", e);
```

Why are we catching Exception and mapping to QueueException?

Only DataStoreException can be thrown from that block, so it seems like we should let that propagate out to handlePublish and handle it there.

---

done